### PR TITLE
add support for named conda environments

### PR DESCRIPTION
**Pull request by nh13**
Use the conda_env directive

Implements: https://github.com/snakemake/snakemake/issues/352

### CODE REVIEW

File: snakemake/parser.py [hunk 1]
```diff
@@ -427,6 +427,12 @@ class Conda(RuleKeywordState):
     pass
 
 
+class CondaEnv(RuleKeywordState):
+    @property
+    def keyword(self):
+        return "conda_env"
```
Comment (nh13) [inline snakemake/parser.py:433]: this can be `condaenv` but using snakecase to make it clearer

File: snakemake/utils.py [hunk 2]
```diff
@@ -555,3 +556,9 @@ def _find_bash_on_windows():
         except FileNotFoundError:
             bashcmd = ""
     return bashcmd if os.path.exists(bashcmd) else None
+
+
+CondaEnvTuple = namedtuple(
+    "CondaEnvTuple",
+    "is_named env"
+)
```
Comment (nh13) [inline snakemake/utils.py:564]: This could also be removed, and instead of using `is_named`, we just using `isinstance(conda_env, str)` in its stead.

**Comment by johanneskoester** [comment]
Thank you! I'll try to review this ASAP. Just give me some additional time as there is a deadline tomorrow and I need to release some urgent fixes in Varlociraptor.

**Comment by johanneskoester** [comment]
Further, may I ask you to format the PR with black?

**Comment by nh13** [comment]
@johanneskoester I'll try to disentangle the reliance on an environment file existing, as that's what I was trying to make work.  Likely I'll get to this in early September when I have some time off for side-projects :)

**Comment by conchoecia** [comment]
I take it that this still isn't resolved? Just was trying to figure out calling up a conda env for a single rule today!

**Comment by nh13** [comment]
@conchoecia I haven't had time to get around to this, and have stuck with boiler plate code for now.

I like to store all my conda environments in one place using [`conda-env-builder`](https://github.com/nh13/conda-env-builder) (which I wrote).  It's super handy if I have a single git repo where I'm building multiple docker images (each a snakemake pipeline), each with one or more conda environments.  

In the snakefile I use the following shell boilerplate:
```smk
rule:    
    params:
        conda_env = "some_prebuilt_and_named_env"
    shell:
        """
        set +eu
        && PS1=dummy
        && . $(conda info --base)/etc/profile.d/conda.sh
        && conda activate {params.conda_env};
        set -eu 
        ...
        """

**Comment by nh13** [comment]
@johanneskoester I really tried to reason about how `conda_env` was really used in `snakemake`. So much of the code is predicated on having a valid and present environments YAML file, I had a lot of trouble finding every place it is used (jobs, persistence, cleanup), and coming up with a good abstraction.  So many objects (workflows, rules, persistence) keep copies of things, and it's really hard to tell exactly what (path, `Env` class, string) without types.  Sorry for the rant, I do think that it could really benefit from some refactoring like you suggested, but that's beyond the time I have right now.

I think the latest set of changes are good compromise for a minimal set of changes to get it working.  Thank-you for your consideration!

**Comment by nh13** [comment]
@johanneskoester tests pass, this is ready to go!

**Comment by mw55309** [comment]
Just adding a +1 for this feature - is there a timeline for implementation?

**Comment by LiterallyUniqueLogin** [comment]
I'm also interested in this. I also don't need snakemake to manage my conda environments for me. Given that the code has already been helpfully written, can we get traction on this?

**Comment by colinbrislawn** [comment]
Any updates?

I'm looking forward to using Snakemake with complex (but standardized!) environments like this one:
https://data.qiime2.org/distro/core/qiime2-2021.11-py38-linux-conda.yml

**Comment by snystrom** [comment]
Chiming in to say this would be very useful in a couple of my more esoteric workflows with external non-conda deps.

**Comment by Kdreval** [comment]
This would be useful enhancement of snakemake functionality! Any updates on the timeline for this feature?

**Review by johanneskoester — COMMENTED** [review]
Thanks a lot for pointing me again to this. Sorry for the delay. This slipped my attention, because, to be honest, I personally rather favor to have the yaml in the workflow definition, because I think this is much more transparent and adaptable. Nevertheless, Snakemake best practices are not a religion and if we can make a change that fits your habits without totally rewriting the way Snakemake works, I am fine with giving it a try.

Nevertheless, I don't want to merge this in the current form, because of the hack using something considered to be a file path as a name instead. I really think the OO solution is worth giving a try. But I know, this is quite some refactoring work, and I can understand that you probably also don't have the time for this. Given the fact that I forgot this PR for such a long time and you patiently waited for it, I will have a look into this today. Maybe I can come up with a solution quickly.

**Comment by nh13** [comment]
Thank you @johanneskoester for giving it a shot. I commend you for making reproducibility easy and the first choice when running Snakemake, but also understand the pragmatic choices that other people (including myself) make that require alternatives. Like you said, we shouldn’t become dogmatic about the right way to ensure reproducibility. In some cases, I want Snakemake just to the excellent workflow engine that it is, and not force me down a specific path to build my environment. I’d argue that it’s great both are included in Snakemake, but should be allowed to be decoupled if wanted. 

I did try to figure out how to implement the OO approach, but like I said, there’s a lot of coupling in the code and as a newcomer to the code it was quite daunting. I’d love to learn and contribute more (along with our team) as we really love Snakemake and use it in our day to day across a large set of clients. 

Seeing how you’d approach this feature could help us understand the code, and later we can add more functionality and/or code docs to help our future selves.  Thank you for your consideration!

**Comment by colinbrislawn** [comment]
Thanks for returning to this PR, @johanneskoester

I'm glad we waited to revisit this issue. Much has changed, so my use case for named environments is different.

My user story: most rules I write call a single program (bwa) or some unix tools (grep + sed). I find it easy to
1) write env.yaml files and --use-conda or
2) call snakemake from within whatever conda environment I am building

But sometimes I use Qiime2. Each version has +400 dependencies, and is installed with conda into an environment with a [fixed name](https://docs.qiime2.org/2021.11/install/native/#install-qiime-2-within-a-conda-environment). Because the conda environment for each release of is consistent, calling `conda activate qiime2-2021.11` should work for anyone already using Qiime2.

Before mamba, installing more copies of qiime2 was very slow. But times have changed! I'll try putting this into an snakemake conda env and see how it goes!

**Comment by colinbrislawn** [comment]
Well, [qiime2-2021.11-py38](https://data.qiime2.org/distro/core/qiime2-2021.11-py38-linux-conda.yml) is causing it's own issues

<details>
  <summary>Error with qiime2-2021.11-py38</summary>
  
Creating conda environment workflow/envs/qiime2-2021.11-py38-linux-conda.yml...
Downloading and installing remote packages.
CreateCondaEnvironmentException:
Could not create conda environment from /net/fs01.cluster.com/home/cbrislawn/bin/github-repos/unite-train/workflow/envs/qiime2-2021.11-py38-linux-conda.yml:
Encountered problems while solving.
Problem: nothing provides __glibc >=2.17,<3.0.a0 needed by nodejs-16.12.0-h73e1104_0
Problem: nothing provides __glibc >=2.17,<3.0.a0 needed by nodejs-16.12.0-h73e1104_0 

</details>

For some reason `conda` is able to solve this environment, mamba can not... 🧐

I suppose this shows another use case for named environments.

**Pull request closed without merging**